### ISSUE TITLE
Excluding children and nested shared-elements

### ISSUE THREAD

**Issue by Chaphasilor** [justintaddei/v-shared-element#18]
I have a shared element with `includeChildren: true`. This transitions a small card into a larger, more detailed one.
On the larger card, some elements exist that are also on the smaller card, but not all.

So I'm trying to transition the smaller card into the larger one, but also having shared elements *inside* the smaller card that have a matching element in the larger card.  
The problem I'm facing is that I can't exclude these inner shared elements, causing illusory to transition them twice, and to different locations. Here's a video demo:

https://streamable.com/6ucfrl

If you look at the progress bar (the gray one), you can see how two copies of it are created.  
Is it somehow possible to prevent this? I know you can filter out elements in Illusory, but the `processClone` option doesn't seem to be available in the `v-shared-element`-options...

I think excluding some would be a far better option than manually transitioning every single element using `includeChildren: false`...

**Comment by justintaddei** [comment]
Would it possible for you to share the markup for the two cards? If I understand correctly, what you've done should've worked as you intended.

My initial thought is that this is a bug, but if it's not, it's definitely a feature worth adding.

Perhaps `v-shared-element.exclude`?

**Comment by Chaphasilor** [comment]
sure, here's the relevant markup:

smaller card
```vue
<div
  :class="`mx-4 my-2 bg-white rounded-xl shadow-md border flex flex-row flex-wrap justify-between p-5 leading-9`"
>

  <div
    :class="(!progressBarStates.includes(download.status) ? 'w-3/4' : 'w-full') + ` text-left font-semibold truncate h-8`"
    v-shared-element:[`${download.id}-title`]="{
      zIndex: 2,
    }"
  >
    {{ download.filename }}
  </div>

  <div
    v-if="showProgressbar"
    class="w-7/8 h-8 flex flex-col justify-center"
    v-shared-element:progress-bar="{
      zIndex: 3,
    }"
  >
    <ProgressBar
      class="w-full h-2 overflow-hidden"
      :percentage="download.percentage"
    />
  </div>

  <div
    :class="`${(!progressBarStates.includes(download.status) ? `w-1/4` : `w-1/8`)} text-right h-8 ${download.textColors.red.includes(download.status) ? `text-cta-red` : download.textColors.yellow.includes(download.status) ? `text-cta-yellow` : download.textColors.green.includes(download.status) ? `text-cta-green` : `text-dark` }`"
    v-shared-element:[`${download.id}-status`]="{
      zIndex: 2,
    }"
  >
    {{ statusString }}
  </div>

</div>
```

larger card
```vue
<div
  class="h-screen"
>
  <div
    class="w-full h-16 p-4 flex flex-row justify-between"
  >

    <h2
      class="w-5/6 flex-shrink-0 text-left text-2xl antialiased font-semibold text-dark tracking-wide"
      v-shared-element:page-title    
    >
      Download Details
    </h2>

    <div
      v-shared-element:settings-icon
      class="w-8 h-8 text-dark stroke-current stroke-1.5 active:stroke-2 transition-colors duration-300"
    >
    <!-- Dummy element used for fading out the settings icon -->
    </div>

  </div>

  <div
    class="mx-4 my-2 bg-white text-dark rounded-xl shadow-md p-5 leading-9"
    v-shared-element:[download.id]
  >

    <div
      class="w-full text-left mb-2 font-semibold break-all h-8 overflow-hidden"
      v-shared-element:[`${download.id}-title`]="{
        zIndex: 2,
      }"
    >
      {{ download.filename }}
    </div>

    <div
      v-if="showProgressbar"
      class="w-full h-8 my-2 flex flex-row flex-wrap justify-between"
      v-shared-element:progress-bar="{
        zIndex: 3,
      }"
    >
      <ProgressBar
        class="w-full h-6 overflow-hidden"
        :percentage="download.percentage"
      />
    </div>

    <InfoLine
      class="w-full h-8"
      name="Status"
      :value="statusString"
      :color="download.textColors.red.includes(download.status) ? `text-cta-red` : download.textColors.yellow.includes(download.status) ? `text-cta-yellow` : download.textColors.green.includes(download.status) ? `text-cta-green` : `text-dark`"
      v-shared-element:[`${download.id}-status`]="{
        zIndex: 2,
      }"
    />

    <!-- more items that are not on the smaller card -->

    <router-link
      class="w-full"
      :to="{
        name: 'Progress',
      }"
    >
      <CTAButton
        class="w-full h-12 mt-6"
        type="good"
        label="Back"
      />
    </router-link>
  </div>
</div>
```

**Comment by Chaphasilor** [comment]
I agree that this would be a feature worth adding, if there's no way to do this yet :)

**Comment by justintaddei** [comment]
Is the smaller card missing `v-shared-element:[download.id]`?
```vue
<div
  :class="`mx-4 my-2 bg-white rounded-xl shadow-md border flex flex-row flex-wrap justify-between p-5 leading-9`"
>
```
vs
```vue
<div
  :class="`mx-4 my-2 bg-white rounded-xl shadow-md border flex flex-row flex-wrap justify-between p-5 leading-9`"
  v-shared-element:[download.id]
>
```

Another thing that I just noticed is that the smaller and larger cards do no line-up during the animation. Something is definitely going very wrong.

You wouldn't happen to have a live example I could play around with, would you?

**Comment by justintaddei** [comment]
When a shared-element contains another shared-element, it should automatically avoid the duplication issue.

As can be seen in this example:

![test](https://user-images.githubusercontent.com/16705594/95023635-c7d2f980-0643-11eb-9127-112cf51c67e4.gif)

I currently have no clue why it isn't working in your case, but I'll keep trying :)

**Comment by Chaphasilor** [comment]
> Is the smaller card missing `v-shared-element:[download.id]`?

the directive is on the custom component:

```vue
<ProgressCard
  :key="item.id"
  v-for="item of allItems"
  :download="item"
  v-shared-element:[item.id]
/>
```

> Another thing that I just noticed is that the smaller and larger cards do no line-up during the animation. Something is definitely going very wrong.
I noticed that too, but that's just a minor inconvenience for me...

> You wouldn't happen to have a live example I could play around with, would you?

Not really. You could check out Chaphasilor/njoy-frontend/tree but you'd have to modify the source a bit...  
I could provide you with a working preview later if you need it?

**Comment by Chaphasilor** [comment]
> Edit: What browser(s) have you tested and is the same thing happening in other browsers?

Tested on Edge Chromium Dev (latest version) and Firefox Dev (82.0b6)

**Comment by justintaddei** [comment]
> the directive is on the custom component:  

Ah okay, that makes more sense.


Building from your repo should work fine. I'll let you know if I have any trouble.

If that's happening on chromium then I'm confident it's not the browser's fault (i've had issues with Safari in the past).

I'll let you know when I figure something out.

**Comment by Chaphasilor** [comment]
Give me a minute to set up a dummy example :)

**Comment by Chaphasilor** [comment]
Should be working now. Let me know if you have any questions! :D

**Comment by justintaddei** [comment]
Will do.

Thank you!

**Comment by justintaddei** [comment]
**TL;DR**
`visibility: hidden;` is being reversed on child elements, making them visible again. I'll either release a patch later this afternoon or tomorrow morning as time permits.


Who knew that `visibility: hidden;` can be revered on child elements to make them visible again? I didn't . . .  
. . . I guess you learn something new every day.
If you're curious:
https://css-tricks.com/snippets/css/toggle-visibility-when-hiding-elements/


What happened is that when illusory copies an element it applies all of it's computed styles, including `visibility` which defaults to `visible`.

I'm using `visibility: hidden;` to hide the duplicate elements v-shared-element:
```ts
// . . .
    processClone(node, depth) {
      // Hide any nested shared elements if they're currently animating
      // because otherwise they'll show up in the clone and in their
      // animation
      if (
        depth > 0 &&
        (node instanceof HTMLElement || node instanceof SVGElement) &&
        node.dataset.illusoryId &&
        sharedElementCache.has(node.dataset.illusoryId)
      )
        node.style.visibility = 'hidden'
    },
// . . .
```
https://github.com/justintaddei/v-shared-element/blob/7de35e9ca21d683a512c586b759c1fcfe5facad7/src/index.ts#L55 

That works fine unless, as in your case, you have elements nested more than 1 node deep (apparently I never tested this 😬).  

What I'm going to do is replace `visibility` with `opacity`.

**Comment by justintaddei** [comment]
>> Another thing that I just noticed is that the smaller and larger cards do no line-up during the animation. Something is definitely going very wrong.  

>I noticed that too, but that's just a minor inconvenience for me...

The reason this is happening that the shared-element is not the actual card that you see, rather the `div` that is wrapping it.

If you make `<BetterRouterLink>` the root-node of `<ProgressCard>` it will fix the issue.

```vue
<template>
  <BetterRouterLink
    :to="{
      name: 'DownloadDetails',
      params: {
        downloadId: download.id,
        download: download,
      },
    }"
    :class="`mx-4 my-2 bg-white rounded-xl shadow-md border flex flex-row flex-wrap justify-between p-5 leading-9`"
  >

<!-- instead of -->

<template>
  <div>
    <BetterRouterLink
      :to="{
        name: 'DownloadDetails',
        params: {
          downloadId: download.id,
          download: download
        }
      }"
      :class="`mx-4 my-2 bg-white rounded-xl shadow-md border flex flex-row flex-wrap justify-between p-5 leading-9`"
    >
```

**Comment by Chaphasilor** [comment]
> The reason this is happening that the shared-element is not the actual card that you see, rather the `div` that is wrapping it.
> 
> If you make `<BetterRouterLink>` the root-node of `<ProgressCard>` it will fix the issue.

Thanks for the heads up! Any general rule of thumb for which element to use the directive on?  
I also had problems with elements being transitioned to the wrong position and glitching back once the transition is finished.

I'll see if I can set up another example...

Oh and btw, have you seen my issues in https://github.com/justintaddei/vue-haptic/issues? \^\^

**Comment by justintaddei** [comment]
Interesting. I'll have to find a way to overcome inline elements.

As for general rules, only use the directive on the "visual element." That is, not on wrapper elements. The reason is that often wrapper elements have an aspect ratio different from that of the element they are wrapping. Also, border-radius will not be properly accounted for as you saw in your app. Someday I'll write a best-practices section in the docs.


> Oh and btw, have you seen my issues in https://github.com/justintaddei/vue-haptic/issues? ^^

No, I haven't. For whatever reason, I didn't get the notification. I'll look at it right now!

**Comment by Chaphasilor** [comment]
Awesome thanks you so much!

**Comment by Chaphasilor** [comment]
Just updated the packages and IT LOOKS AWESOME!  
Thanks again!

**Comment by justintaddei** [comment]
Thank you for bringing it to my attention!
